ceno_zkvm/src/e2e.rs

```diff
@@ -219,38 +142,257 @@ pub fn run_e2e_gen_witness<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>
+        &system_config.zkvm_cs,
+        &mut zkvm_fixed_traces,
+        &init_mem_state.reg,
+        &init_mem_state.mem,
+        &init_mem_state.io.iter().map(|rec| rec.addr).collect_vec(),
+    );
+    system_config
+        .dummy_config
+        .generate_fixed_traces(&system_config.zkvm_cs, &mut zkvm_fixed_traces);
+
+    zkvm_fixed_traces
+}
+
+fn generate_witness<E: ExtensionField>(
+    system_config: &ConstraintSystemConfig<E>,
+    sim_result: SimulationResult,
+    program: &Program,
+) -> ZKVMWitnesses<E> {
+    let mut zkvm_witness = ZKVMWitnesses::default();
+    // assign opcode circuits
+    let dummy_records = system_config
+        .config
+        .assign_opcode_circuit(
+            &system_config.zkvm_cs,
+            &mut zkvm_witness,
+            sim_result.all_records,
+        )
+        .unwrap();
+    system_config
+        .dummy_config
+        .assign_opcode_circuit(&system_config.zkvm_cs, &mut zkvm_witness, dummy_records)
+        .unwrap();
+    zkvm_witness.finalize_lk_multiplicities();
+
     // assign table circuits
-    config
-        .assign_table_circuit(&zkvm_cs, &mut zkvm_witness)
+    system_config
+        .config
+        .assign_table_circuit(&system_config.zkvm_cs, &mut zkvm_witness)
         .unwrap();
-    mmu_config
+    system_config
+        .mmu_config
         .assign_table_circuit(
-            &zkvm_cs,
+            &system_config.zkvm_cs,
             &mut zkvm_witness,
-            &reg_final,
-            &mem_final,
-            &io_final,
-            &priv_io_final,
+            &sim_result.final_mem_state.reg,
+            &sim_result.final_mem_state.mem,
+            &sim_result
+                .final_mem_state
+                .io
+                .iter()
+                .map(|rec| rec.cycle)
+                .collect_vec(),
+            &sim_result.final_mem_state.priv_io,
         )
         .unwrap();
     // assign program circuit
     zkvm_witness
-        .assign_table_circuit::<ProgramTableCircuit<E>>(&zkvm_cs, &prog_config, vm.program())
+        .assign_table_circuit::<ProgramTableCircuit<E>>(
+            &system_config.zkvm_cs,
+            &system_config.prog_config,
+            program,
+        )
         .unwrap();
 
+    zkvm_witness
+}
+
+pub enum PipelinePrefix {
+    PreProving,
+    PreVerifying,
+    Complete,
+}
+
+pub type ProvingArgs<E, PCS> = (ZKVMProver<E, PCS>, ZKVMWitnesses<E>, PublicValues<u32>);
+pub type VerifyingArgs<E, PCS> = (ZKVMProof<E, PCS>, ZKVMVerifier<E, PCS>, Option<u32>);
+
+pub enum PipelineResult<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> {
+    PreProving(ProvingArgs<E, PCS>),
+    PreVerifying(VerifyingArgs<E, PCS>),
+    Complete,
+}
+
+pub fn run_partial<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>>(
+    program: Program,
+    platform: Platform,
+    stack_size: u32,
+    heap_size: u32,
+    hints: Vec<u32>,
+    max_steps: usize,
+    prefix: PipelinePrefix,
```

Comment:
It seems that prefix can be Option, and without giving, it's run as a whole.
Probably better naming of function is `run_e2e_with_checkpoints` :) ?

Reply:
Agreed with name change. 

It seems more natural to have a `Complete` option for the `Checkpoint` enum, because the function needs to return the `CheckpointOutput` type anyway, and these two enums look nicer with identical variants.